src/main/java/org/traccar/web/server/reports/ReportGraph.java

```diff
@@ -0,0 +1,185 @@
+package org.traccar.web.server.reports;
+
+import org.traccar.web.shared.model.*;
+
+import java.text.SimpleDateFormat;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ReportGraph extends ReportGenerator {
+    @Override
+    void generateImpl(Report report) throws IOException {
+        h2(report.getName());
+
+        //     List<Position> position = entityManager.createQuery("SELECT p FROM Position p WHERE p.device = :device AND p.speed > 0 ORDER BY time DESC", Position.class)
+
+        List<Position> positions = entityManager.createQuery("SELECT p FROM Position p" +
+                " WHERE p.device IN :selectedDevices AND p.time BETWEEN :from AND :to ORDER BY p.time", Position.class)
+                .setParameter("selectedDevices", getDevices(report))
+                .setParameter("from", report.getFromDate())
+                .setParameter("to", report.getToDate())
+                .getResultList();
+
+
+        panelStart();
+
+        // heading
+        panelHeadingStart();
+        text("Graph ...");
+        panelHeadingEnd();
+
+        // body
+        panelBodyStart();
+        // period
+        paragraphStart();
+        bold(message("timePeriod") + ": ");
+        text(formatDate(report.getFromDate()) + " - " + formatDate(report.getToDate()));
+        paragraphEnd();
+
+        try {
+            addGraph(getDevices(report), positions);
+        } catch (TraccarException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
+        }
+
+
+        panelBodyEnd();
+
+        panelEnd();
+
+
+    }
+
+    static class Data {
+        final boolean idle;
+
+        Position start;
+        Position end;
+        double topSpeed;
+        double totalSpeed;
+        double distance;
+        int positionsWithSpeed;
+
+        Data(boolean idle, Position start) {
+            this.start = start;
+            this.idle = idle;
+        }
+
+        long getDuration() {
+            return end.getTime().getTime() - start.getTime().getTime();
+        }
+
+        double getAverageSpeed() {
+            return totalSpeed / positionsWithSpeed;
+        }
+    }
+
+
+
+
+    public void addGraph(List<Device> list, List<Position> positions) throws TraccarException{
+
+        text("<div id=\"graphdiv\" style=\"width:1100px; height:400px;\"></div>");
```

Comment:
You should never put HTML code in report calculator. It should be put into renderer. The main idea of reports design was to separate generator from the renderer. This design will allow rendering of reports into different formats other than HTML.